File: cpp/3. Longest Substring Without Repeating Characters.cpp
/*

2016-11-6 23:19:37

https://leetcode.com/problems/longest-substring-without-repeating-characters/


Given a string, find the length of the longest substring without repeating characters.

Examples:

Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.


*/


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
    	unordered_map<char, int> mymap;

    	int res = 0;
    	int tmpRes = 0;
    	int threIndex = 0;

    	const char* str = s.c_str();
    	for (int i = 0, len = s.size(); i != len; ++ i)
    	{
    		if (mymap.find(str[i]) == mymap.end())
    		{
    			++ tmpRes;
    			mymap[str[i]] = i;
    		}
    		else if (mymap[str[i]] < threIndex)
    		{
    			++ tmpRes;
    			mymap[str[i]] = i;
    		}
    		else
    		{
    			if (tmpRes > res)
    			{
    				 res = tmpRes;
    			}
    			threIndex = mymap[str[i]] + 1;
    			tmpRes = i - mymap[str[i]];
    			mymap[str[i]] = i;
    		}
    	}

    	if (tmpRes > res)
    	{
    		 res = tmpRes;
    	}

    	return res;

    }
};

File: java/3. Longest Substring Without Repeating Characters.java
//2018-05-01 14:37:41
//答案里，如果假设字符集大小128个，可以直接创建128大小的数组解决。

class Solution {
    public int lengthOfLongestSubstring(String s) {

        if (s == null)
            return 0;

        int tail = 0;

        int res = 0;

        HashMap<Character, Integer> hm = new HashMap<>();

        for (int i = 0; i != s.length(); ++ i) {
            if (hm.containsKey(s.charAt(i)) && hm.get(s.charAt(i)) >= tail) {
                tail = hm.get(s.charAt(i)) + 1;
            }

            hm.put(s.charAt(i), i);


            if (res < i - tail + 1)
                res = i - tail + 1;
            
            //System.out.println(res);

        }

        return res;
        
    }
}